src/java.base/share/classes/jdk/internal/classfile/ClassReader.java

```diff
@@ -0,0 +1,270 @@
+    int thisClassPos();
+
+    /** {@return the length of the classfile, in bytes} */
+    int classfileLength();
+
+    // Buffer related
+
+    /**
+     * {@return the offset following the block of attributes starting at the
+     * specified position}
+     * @param offset the offset into the classfile at which the attribute block
+     *               starts
+     */
+    int skipAttributeHolder(int offset);
+
+    // Constant pool
+
+    /**
+     * {@return the UTF8 constant pool entry at the given index of the constant
+     * pool}  The given index must correspond to a valid constant pool index
+     * whose slot holds a UTF8 constant.
+     * @param index the index into the constant pool
+     */
+    Utf8Entry utf8EntryByIndex(int index);
+
+    /**
+     * {@return the constant pool entry whose index is given at the specified
+     * offset within the classfile}
+     * @param offset the offset of the index within the classfile
+     * @throws IndexOutOfBoundsException if the index is out of range of the
+     *         constant pool size, or zero
+     */
+    PoolEntry readEntry(int offset);
+
+    /**
+     * {@return the constant pool entry whose index is given at the specified
+     * offset within the classfile, or null if the index at the specified
+     * offset is zero}
+     * @param offset the offset of the index within the classfile
+     * @throws IndexOutOfBoundsException if the index is out of range of the
+     *         constant pool size
+     */
+    PoolEntry readEntryOrNull(int offset);
+
+    /**
+     * {@return the UTF8 entry whose index is given at the specified
+     * offset within the classfile}
+     * @param offset the offset of the index within the classfile
+     * @throws IndexOutOfBoundsException if the index is out of range of the
+     *         constant pool size, or zero
+     * @throws IllegalArgumentException if the index does not correspond to
+     *         a UTF8 entry
+     */
+    Utf8Entry readUtf8Entry(int offset);
+
+    /**
+     * {@return the UTF8 entry whose index is given at the specified
+     * offset within the classfile, or null if the index at the specified
+     * offset is zero}
+     * @param offset the offset of the index within the classfile
+     * @throws IndexOutOfBoundsException if the index is out of range of the
+     *         constant pool size
+     * @throws IllegalArgumentException if the index does not correspond to
+     *         a UTF8 entry
+     */
+    Utf8Entry readUtf8EntryOrNull(int offset);
+
+    /**
+     * {@return the module entry whose index is given at the specified
+     * offset within the classfile}
+     * @param offset the offset of the index within the classfile
+     * @throws IndexOutOfBoundsException if the index is out of range of the
+     *         constant pool size, or zero
+     * @throws ClassCastException if the index does not correspond to
+     *         a module entry
```

Comment:
Throwing `ClassCastException` is certainly convenient from an implementation perspective, but I think it is better to throw `IllegalArgumentException`.

Reply:
Fixed in all `ClassReader::readXyzEntry` methods, thanks.